Sources/WebSocketKit/Concurrency/WebSocket+Concurrency.swift

```diff
@@ -0,0 +1,145 @@
+#if compiler(>=5.5) && canImport(_Concurrency)
+import NIOCore
+import NIOWebSocket
+import Foundation
+
+@available(macOS 12, iOS 15, watchOS 8, tvOS 15, *)
+extension WebSocket {
+    public func send<S>(_ text: S) async throws
+        where S: Collection, S.Element == Character
+    {
+        let promise = eventLoop.makePromise(of: Void.self)
+        send(text, promise: promise)
+        return try await promise.futureResult.get()
+    }
+
+    public func send(_ binary: [UInt8]) async throws {
+        let promise = eventLoop.makePromise(of: Void.self)
+        send(binary, promise: promise)
+        return try await promise.futureResult.get()
+    }
+
+    public func sendPing() async throws {
+        let promise = eventLoop.makePromise(of: Void.self)
+        sendPing(promise: promise)
+        return try await promise.futureResult.get()
+    }
+
+    public func send<Data>(
+        raw data: Data,
+        opcode: WebSocketOpcode,
+        fin: Bool = true
+    ) async throws
+        where Data: DataProtocol
+    {
+        let promise = eventLoop.makePromise(of: Void.self)
+        send(raw: data, opcode: opcode, fin: fin, promise: promise)
+        return try await promise.futureResult.get()
+    }
+
+    public func close(code: WebSocketErrorCode = .goingAway) async throws {
+        try await close(code: code).get()
+    }
+
+    public func onText(_ callback: @escaping (WebSocket, String) async -> ()) {
+        onText { socket, text in
+            Task {
```

Comment:
For some reason this is making me uneasy but I think this is the correct way. Concur @gwynne ?

Reply:
@0xTim Sorry for jumping in, but I think it's wrong to be like this, for the following reason:

```Swift
    Task {
      do {
        try await WebSocket.connect(to: URL(string: "wss://socketURL")!, on: eventLoopGroup) { ws in
          
          // Does not work and it's very easy to get confused with!
          ws.onText({ ws, text in
            print(text)
          })
          
          // Works but gives a warning
          // No 'async' operations occur within 'await' expression
          await ws.onText({ ws, text in
            print(text)
          })
        }
      } catch {
        print(error)
      }
    }
```

Shouldn't it be something like this?
```Swift
let (ws, text) = await ws.getText()

```

Reply:
In this case no, a closure is the right approach because it's invoked each time text is sent through the web socket channel

Reply:
I get it but how to solve these to problems?
* You forget (there are not enforcement) to use await before the closure, and if you forget it will not work at all.
* How to get around/silence the Xcode warning?